README.txt
=============
CPSUid README
=============

:Author: Anahide Tchertchian
:Revision: $Id$


CPSUid provides a unique identifier implementation for CPS documents.


Features
========

CPSUid provides :

- identification profiles registration
- unique id registration: unique ids are stored on the object, and
  incremented given the last unique id used.
- unique id handling: given a uid, objects using this id can be found.


Install
=======

CPSUid comes with an extension profile definition.
Go into portal_setup at the root of the site, choose the 'import' tab, and
import the CPSUid profile.



More information
================

See the doc subsection



.. Emacs
.. Local Variables:
.. mode: rst
.. End:
.. Vim
.. vim: set filetype=rst:


CHANGES
Requires
~~~~~~~~
-
New features
~~~~~~~~~~~~
-
Bug fixes
~~~~~~~~~
- Fixed unit tests that had been broken for 4 years
New internal features
~~~~~~~~~~~~~~~~~~~~~
- 


tests/test_uidtool.py
# Copyright (c) 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""Tests for the uid generator module

$Id$
"""

import unittest

from zope.interface.verify import verifyClass

from Products.CPSUid.interfaces import IUidGenerator
from Products.CPSUid.uidgenerator import UidGenerator

class TestUidGenerator(unittest.TestCase):

    # fixture

    def setUp(self):
        # add a test generator
        self.generation_criteria = ('source',)
        self.generation_keywords = ('portal_type',)
        # TAL expression to build the id as D-S-YY-NNN with D as document type,
        # S as source,YY as the current yea and NNN an incremented number
        self.expression = "python:str(portal_type)+'-'+str(source)+'-'+DateTime().strftime('%y')+'-'+'%#03d'%number"
        self.generator = UidGenerator(
            'generator',
            generation_criteria=self.generation_criteria,
            generation_keywords=self.generation_keywords,
            generation_expression=self.expression)

    # tests

    def test_interface(self):
        verifyClass(IUidGenerator, UidGenerator)

    def test_creation(self):
        expression = "python:str(foo)+'_'+str(bar)"
        generator = UidGenerator('generator', generation_criteria=('foo'),
                                 generation_keywords=('bar'),
                                 generation_expression=expression)
        self.assertEqual(generator.meta_type, "Uid Generator")
        self.assertEqual(generator.getId(), 'generator')
        self.assertEqual(generator.generation_criteria, ('foo',))
        self.assertEqual(generator.generation_keywords, ('bar',))
        self.assertEqual(generator.generation_expression, expression)
        self.assertNotEqual(generator.generation_expression_c, None)

    def test_test_case_generator(self):
        self.assertEqual(self.generator.meta_type, "Uid Generator")
        self.assertEqual(self.generator.getId(), 'generator')
        self.assertEqual(self.generator.generation_criteria,
                         self.generation_criteria)
        self.assertEqual(self.generator.generation_keywords,
                         self.generation_keywords)
        self.assertEqual(self.generator.generation_expression,
                         self.expression)
        self.assertNotEqual(self.generator.generation_expression_c, None)

    def test_getUid(self):
        kw = {
            'portal_type': 'file_document',
            'source': 'CPS',
            }
        uid = self.generator.getUid(**kw)
        self.assertEqual(uid, 'file_document-CPS-06-000')
        new_uid = self.generator.getUid(**kw)
        self.assertEqual(new_uid, 'file_document-CPS-06-001')
        # hit again
        newer_uid = self.generator.getUid(**kw)
        self.assertEqual(newer_uid, 'file_document-CPS-06-002')


    def test__getCounter(self):
        criteria = {
            'source': 'CPS',
            }
        counter = self.generator._getCounter(**criteria)
        self.assertEqual(counter.getId(), 'source-CPS')
        self.assertEqual(counter.counter_start, 0)
        self.assertEqual(counter.counter_current, 0)
        self.assertEqual(counter.criteria, criteria)


    def test___createCounter(self):
        criteria = {
            'source': 'CPS',
            }
        counter = self.generator._createCounter(**criteria)
        self.assertEqual(counter.getId(), 'source-CPS')
        self.assertEqual(counter.counter_start, 0)
        self.assertEqual(counter.counter_current, 0)
        self.assertEqual(counter.criteria, criteria)


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(TestUidGenerator))
    return suite


tests/test_uidcounter.py
# Copyright (c) 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""Tests for the uid counter module

$Id$
"""

import unittest
from Testing.ZopeTestCase import ZopeTestCase

from zope.interface.verify import verifyClass

import transaction
from Products.CPSUid.interfaces import IUidCounter
from Products.CPSUid.uidcounter import UidCounter

class TestUidCounter(ZopeTestCase):

    # fixture

    def afterSetUp(self):
        # add a test counter
        self.criteria = {
            "source": "CPS",
            "doument_type": "Common type",
            }
        self.criteria_lines = (
            "source CPS",
            "doument_type Common type",
            )
        self.folder._setObject('counter',
                               UidCounter("counter", 1, self.criteria))
        self.counter = self.folder.counter
        transaction.commit()

    def beforeTearDown(self):
        self.app._delObject(self.folder.getId())
        transaction.commit()

    # tests

    def test_interface(self):
        verifyClass(IUidCounter, UidCounter)

    def test_creation(self):
        criteria = {
            "source": "CPS",
            "doument_type": "Common type",
            }
        counter = UidCounter("counter", 1, criteria)
        self.assertEqual(counter.meta_type, "Uid Counter")
        self.assertEqual(counter.getId(), "counter")
        self.assertEqual(counter.counter_start, 1)
        self.assertEqual(counter.counter_current, 1)
        self.assertEqual(counter.getCriteria(), criteria)

    def test_test_case_counter(self):
        self.assertEqual(self.counter.meta_type, "Uid Counter")
        self.assertEqual(self.counter.getId(), "counter")
        self.assertEqual(self.counter.counter_start, 1)
        self.assertEqual(self.counter.counter_current, 1)
        self.assertEqual(self.counter.criteria, self.criteria_lines)

    def test_hit(self):
        self.assertEqual(self.counter.counter_current, 1)
        self.assertEqual(self.counter.hit(), 1)
        self.assertEqual(self.counter.hit(), 2)
        # all these hits actually belong to the micro-transaction
        # we need to close this one to read the proper value if we want
        # to do it directly
        transaction.commit()
        self.assertEqual(self.counter.counter_current, 3)

    def test_reset(self):
        self.assertEqual(self.counter.counter_current, 1)
        self.counter.counter_current = 3
        transaction.commit()
        self.assertEqual(self.counter.counter_current, 3)
        self.counter.reset()
        transaction.commit()
        self.assertEqual(self.counter.counter_current, 1)

    def test__getCriteria(self):
        self.assertEqual(self.counter.getCriteria(), self.criteria)


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(TestUidCounter))
    return suite


tests/test_exportimport.py
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id$
"""Tests for the CPScore export/import mechanism
"""

import unittest
from Testing import ZopeTestCase

from Products.CPSUtil.testing.genericsetup import ExportImportTestCase

ZopeTestCase.installProduct('CPSUid')

class ExportImportTest(ExportImportTestCase):

    def beforeTearDown(self):
        ExportImportTestCase.beforeTearDown(self)
        self.clearProfileRegistry()


    def test_default_import(self):
        self.registerProfile('default', "CPS Uid", "Default profile",
                             'profiles/default', 'CPSUid')
        self.assertEquals('portal_uid' in self.folder.objectIds(),
                          False)
        self.importProfile('CPSUid:default')

        # check portal_uid
        self.assertEquals('portal_uid' in self.folder.objectIds(),
                          True)


    def test_default_export(self):
        self.registerProfile('default', "CPS Uid", "Default profile",
                             'profiles/default', 'CPSUid')
        self.importProfile('CPSUid:default')
        toc_list = [
            'export_steps.xml',
            'import_steps.xml',
            'uidgenerators.xml',
            ]
        self._checkExportProfile('CPSUid/tests/profiles/default_export/',
                                 toc_list)


    def test_basic_import(self):
        self.registerProfile('basic', "CPS Uid", "Basic profile",
                             'tests/profiles/basic', 'CPSUid')
        self.assertEquals('portal_uid' in self.folder.objectIds(), False)
        self.importProfile('CPSUid:basic')
        self.assertEquals('portal_uid' in self.folder.objectIds(), True)
        uidtool = self.folder.portal_uid
        self.assertEquals(uidtool.meta_type, 'CPS Uid Tool')
        self.assertEquals(list(uidtool.objectIds()), ['generator'])
        generator = uidtool.generator
        self.assertEqual(generator.meta_type, "Uid Generator")
        self.assertEqual(generator.getId(), 'generator')
        self.assertEqual(generator.generation_criteria,
                         ('source',),)
        self.assertEqual(generator.generation_keywords,
                         ('portal_type',))
        # FIXME: import should not keep the extra leadin \n character
        expr = "\npython:str(portal_type)+'-'+str(source)+'-'+DateTime().strftime('%y')+'-'+'%#03d'%number"
        self.assertEqual(generator.generation_expression, expr)
        self.assertNotEqual(generator.generation_expression_c, None)
        self.assertEqual(generator.counter_start, 1)


    def test_basic_export(self):
        self.registerProfile('basic', "CPS Uid", "Basic profile",
                             'tests/profiles/basic', 'CPSUid')
        self.importProfile('CPSUid:basic')
        toc_list = [
            'export_steps.xml',
            'import_steps.xml',
            'uidgenerators.xml',
            'uidgenerators/generator.xml',
           ]
        self._checkExportProfile('CPSUid/tests/profiles/basic',
                                 toc_list)


def test_suite():
    return unittest.TestSuite((
        unittest.makeSuite(ExportImportTest),
        ))



uidtool.py
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""Tool for uid management
"""

from logging import getLogger

from zope.interface import implements

from Globals import InitializeClass
from AccessControl import ClassSecurityInfo
from OFS.Folder import Folder

from Products.CMFCore.utils import UniqueObject
from Products.CMFCore.utils import getToolByName
from Products.CMFCore.permissions import View, ManagePortal

from Products.CPSUid.interfaces import IUidHandler
from Products.CPSUid.uidgenerator import UidGenerator

_marker = []


class UidTool(UniqueObject, Folder):
    """Uid tool to manage document uids
    """

    id = 'portal_uid'
    meta_type = 'CPS Uid Tool'

    implements(IUidHandler)

    security = ClassSecurityInfo()

    logger = getLogger('UidTool')

    # XXX uid setting is done at the field (in schema) level for now

    security.declareProtected(View, 'getUid')
    def getUid(self, generator_id, default=_marker, datamodel=None,
               **kw):
        """Get uid using given generator_id according to keywords
        """
        generator = self._getOb(generator_id, default=None)
        if datamodel is not None:
            kw.update(datamodel.items())
        if generator is not None:
            uid = generator.getUid(**kw)
        else:
            if default is _marker:
                raise AttributeError, generator_id
            else:
                uid = default
        return uid


    # ZMI

    security.declareProtected(ManagePortal, 'manage_addUidGenerator')
    def manage_addUidGenerator(self, id, generation_criteria,
                               generation_keywords, generation_expression,
                               REQUEST=None):
        """Add a uid generator TTW"""
        generator = UidGenerator(id, generation_criteria=generation_criteria,
                                 generation_keywords=generation_keywords,
                                 generation_expression=generation_expression)
        self._setObject(id, generator)
        if REQUEST is not None:
            REQUEST.RESPONSE.redirect(self.absolute_url()+'/manage_main'
                                      '?manage_tabs_message=Generator Added.')

InitializeClass(UidTool)


upgrade.py
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""Upgrades

$Id$
"""

from logging import getLogger
from Products.CMFCore.utils import getToolByName

def checkUidCounters(self):
    upgrade = 0
    uidtool = getToolByName(self, 'portal_uid', None)
    if uidtool is not None:
        for generator in uidtool.objectValues('Uid Generator'):
            for counter in generator.objectValues('Uid Counter'):
                if isinstance(counter.criteria, dict):
                    upgrade = 1
                    break
    return upgrade


def upgradeUidCounters(self):
    """Upgrade uid counters, turning their criteria property into lines
    """
    logger = getLogger('upgradeUidCounters')
    nchanged = 0
    uidtool = getToolByName(self, 'portal_uid', None)
    if uidtool is not None:
        for generator in uidtool.objectValues('Uid Generator'):
            for counter in generator.objectValues('Uid Counter'):
                criteria = counter.criteria
                if isinstance(criteria, dict):
                    logger.debug("upgrading %s" % counter.absolute_url())
                    criteria_lines = list(['%s %s'%(key, value)
                                           for key, value in criteria.items()])
                    counter._setPropValue('criteria', criteria_lines)
                    nchanged += 1
    logger.debug("%s objects upgraded" % nchanged)
    return '%s objects upgraded' % nchanged


def upgrade(self):
    if checkRootWorkflows(self):
        upgradeRootWorkflows(self)


__init__.py
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""Init

$Id$
"""
from Products.CMFCore.utils import ContentInit, ToolInit
from Products.CMFCore.permissions import AddPortalContent, ManagePortal

from Products.GenericSetup import profile_registry
from Products.GenericSetup import EXTENSION
from Products.CPSCore.interfaces import ICPSSite

from Products.CPSUid.uidtool import UidTool
from Products.CPSUid.uidgenerator import UidGenerator

def initialize(registrar):

    ToolInit(
        'CPS Tools',
        tools=(UidTool,),
        icon='tool.png',
        ).initialize(registrar)

    registrar.registerClass(
        UidGenerator,
        permission=ManagePortal,
        constructors=(UidTool.manage_addUidGenerator,),
        )

    profile_registry.registerProfile(
        'default',
        'CPS Uid',
        "Ujnique identifier product for CPS.",
        'profiles/default',
        'CPSUid',
        EXTENSION,
        for_=ICPSSite)


interfaces.py
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""CPSUid interfaces

$Id$
"""

from zope.interface import Interface

class IUidHandler(Interface):
    """Unique identifier handler
    """

    def getUid(generator_id, **kw):
        """Get uid suing given generator_id according to keywords
        """


class IUidGenerator(Interface):
    """Unique identifier generator
    """

    def getUid(**kw):
        """Get uid according to given keywords
        """

    def _getCounter(**criteria):
        """Get counter for given keyword criteria
        """

    def _createCounter(**criteria):
        """Create a counter using given keyword criteria
        """


class IUidCounter(Interface):
    """Unique identifier counter
    """

    def getCriteria():
        """Get the counter criteria
        """

    def hit():
        """Hit the counter, increment its value
        """

    def reset():
        """Reset the counter
        """


uidcounter.py
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
# - Florent Guillaume <fg@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""Unique identifier counter

A counter holds criteria and an integer counter applying to these criteria
"""
import logging

from zope.interface import implements
from transaction import TransactionManager

from Globals import InitializeClass
from AccessControl import ClassSecurityInfo
from ZODB.POSException import ConflictError

from Products.CMFCore.utils import SimpleItemWithProperties

from Products.CPSUid.interfaces import IUidCounter

logger = logging.getLogger('CPSUid.uidcounter')

class UidCounter(SimpleItemWithProperties):
    """Uid Counter

    XXX set persistent right now so that it can hold criteria and counters in
    the ZODB
    """

    implements(IUidCounter)

    meta_type = 'Uid Counter'

    security = ClassSecurityInfo()

    _properties = (
        {'id': 'counter_start', 'type': 'int', 'mode': 'w',
         'label': 'Counter start value'},
        {'id': 'counter_current', 'type': 'int', 'mode': 'w',
         'label': 'Counter current value'},
        # one criteria per line, following the format "key value", for
        # instance:
        # portal_type Workspace
        # type CPS
        {'id': 'criteria', 'type': 'lines', 'mode': 'w',
         'label': 'Criteria mapping'},
        )

    max_tries = 5

    #
    # API
    #

    def __init__(self, id, counter_start=0, criteria={}):
        """Initialization
        """
        self.id = id
        self._setPropValue('counter_start', counter_start)
        self._setPropValue('counter_current', counter_start)
        criteria_lines = list(['%s %s'%(key, value)
                               for key, value in criteria.items()])
        self._setPropValue('criteria', criteria_lines)


    security.declarePrivate('hit')
    def hit(self):
        """Hit the counter: return current counter and increment its value.

        Fires a dedicated micro transaction to diminish risks of conflicts
        because we definitely *don't* want to resolve them (that breaks
        returned values' unicity).

        """
	jar = self._p_jar
	if jar is None:
	   # creation of counter
           v = self.counter_start
           self.counter_current = v + 1
	   return v

        db = self._p_jar.db()

        retries = self.max_tries
        while retries:
            try:
                logger.debug(
                    "Hitting on counter %s, trying %d times", self, retries)
                tm = TransactionManager()

                cnx = db.open(transaction_manager=tm)
                tm.begin()
                ob = cnx.get(self._p_oid)

                v = ob.counter_current
                ob.counter_current = v + 1

                tm.commit()
                cnx.close()
            except ConflictError:
                logger.debug('ConflictError')
                tm.abort()
                cnx.close()
                retries -= 1
            else:
                break
        else:
            # exhausted all tries
            raise ConflictError(object=self)
        return v

    security.declarePrivate('reset')
    def reset(self):
        """Reset the counter
        """
        self.counter_current = self.counter_start


    security.declarePrivate('getCriteria')
    def getCriteria(self):
        """Get the counter criteria
        """
        criteria_dict = {}
        for mapping in self.criteria:
            if ' ' in mapping:
                key, value = mapping.split(' ', 1)
                criteria_dict[key] = value
        return criteria_dict

InitializeClass(UidCounter)


exportimport.py
# Copyright (c) 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""CPSUid XML adapters

$Id$
"""

from zope.component import adapts
from zope.interface import implements

from Products.CMFCore.utils import getToolByName
from Products.GenericSetup.utils import exportObjects
from Products.GenericSetup.utils import importObjects
from Products.GenericSetup.utils import XMLAdapterBase
from Products.GenericSetup.utils import ObjectManagerHelpers

from Products.GenericSetup.interfaces import IBody
from Products.GenericSetup.interfaces import ISetupEnviron

from Products.CPSUtil.PropertiesPostProcessor import \
     PostProcessingPropertyManagerHelpers

from Products.CPSUid.interfaces import IUidHandler
from Products.CPSUid.interfaces import IUidGenerator


UID_TOOL = 'portal_uid'
UID_NAME = 'uidgenerators'


def exportUidTool(context):
    """Export uid tool as a set of XML files.
    """
    site = context.getSite()
    tool = getToolByName(site, UID_TOOL, None)
    if tool is None:
        logger = context.getLogger(UID_NAME)
        logger.info("Nothing to export.")
        return
    exportObjects(tool, '', context)


def importUidTool(context):
    """Import uid tool from a set of XML
    files.
    """
    site = context.getSite()
    tool = getToolByName(site, UID_TOOL)
    importObjects(tool, '', context)


class UidToolXMLAdapter(XMLAdapterBase, ObjectManagerHelpers):
    """XML importer and exporter for Uid tool.
    """
    adapts(IUidHandler, ISetupEnviron)
    implements(IBody)

    _LOGGER_ID = UID_NAME
    name = UID_NAME

    def _exportNode(self):
        """Export the object as a DOM node.
        """
        node = self._getObjectNode('object')
        node.appendChild(self._extractObjects())
        self._logger.info("Uid tool exported.")
        return node

    def _importNode(self, node):
        """Import the object from the DOM node.
        """
        if self.environ.shouldPurge():
            self._purgeObjects()
        self._initObjects(node)
        self._logger.info("Uid tool imported.")


class UidGeneratorXMLAdapter(XMLAdapterBase,
                             PostProcessingPropertyManagerHelpers):
    """XML importer and exporter for a uid generator.
    """
    adapts(IUidGenerator, ISetupEnviron)
    implements(IBody)

    _LOGGER_ID = UID_NAME

    def _exportNode(self):
        """Export the object as a DOM node.
        """
        node = self._getObjectNode('object')
        node.appendChild(self._extractProperties())
        self._logger.info("%s uid generator exported." % self.context.getId())
        return node

    def _importNode(self, node):
        """Import the object from the DOM node.
        """
        if self.environ.shouldPurge():
            self._purgeProperties()
        self._initProperties(node)
        self._logger.info("%s uid generator imported." % self.context.getId())


uidgenerator.py
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Authors:
# - Anahide Tchertchian <at@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
"""Unique identifier generator

Generators hold rules to build identifiers given parameters and also hold
counters.
"""

import random
from logging import getLogger
from DateTime import DateTime

from zope.interface import implements

from Globals import InitializeClass
from AccessControl import ClassSecurityInfo
from OFS.Folder import Folder

from Products.PageTemplates.Expressions import SecureModuleImporter
from Products.CMFCore.permissions import View
from Products.CMFCore.utils import getToolByName
from Products.CMFCore.Expression import getEngine

from Products.CPSUtil.PropertiesPostProcessor import PropertiesPostProcessor
from Products.CPSUtil.id import generateId

from Products.CPSUid.interfaces import IUidGenerator
from Products.CPSUid.uidcounter import UidCounter


def stringify(value):
    if not isinstance(value, basestring):
        value = str(value)
    return value


class UidGenerator(PropertiesPostProcessor, Folder):
    """Uid Generator

    XXX set persistent right now so that it can hold expressions and counters
    in the ZODB
    """

    implements(IUidGenerator)

    meta_type = 'Uid Generator'

    security = ClassSecurityInfo()

    _propertiesBaseClass = Folder
    _properties = (
        {'id': 'generation_expression', 'type': 'string', 'mode': 'w',
         'label': 'Generation expression'},
        {'id': 'generation_criteria', 'type': 'tokens', 'mode': 'w',
         'label': 'Generation criteria'},
        {'id': 'generation_keywords', 'type': 'tokens', 'mode': 'w',
         'label': 'Generation keywords'},
        {'id': 'counter_start', 'type': 'int', 'mode': 'w',
         'label': 'Counter start value'},
        )
    _properties_post_process_tales = (
        ('generation_expression', 'generation_expression_c'),
        )
    generation_expression = ''
    generation_expression_c = None
    generation_criteria = ()
    # XXX generation criteria are generation keywords by default
    generation_keywords = ()
    counter_start = 0

    manage_options = Folder.manage_options

    logger = getLogger("CPSUid.UidGenerator")

    #
    # API
    #

    def __init__(self, id, **kw):
        """Initialization
        """
        self.id = id
        self.manage_changeProperties(**kw)


    security.declareProtected(View, 'getUid')
    def getUid(self, **kw):
        """Get unique identifier value
        """
        if self.generation_expression_c:
            expr = self._createGenerationExpressionContext(**kw)
            uid = self.generation_expression_c(expr)
        else:
            uid = None
        return uid


    security.declarePrivate('_createGenerationExpressionContext')
    def _createGenerationExpressionContext(self, **kw):
        """Create an expression context for generation evaluation.
        """
        # mapping is filled in reverse order so that most important keywords
        # are not overloaded
        mapping = {}
        # generation additional keywords
        for keyword in self.generation_keywords:
            mapping[keyword] = kw.get(keyword)
        mapping['DateTime'] = DateTime
        mapping['modules'] = SecureModuleImporter
        mapping['random'] = random.random
        # counter
        criteria = {}
        # generation criteria
        for keyword in self.generation_criteria:
            # XXX AT: capsule related fix: get rid of colons in keyword name
            # hoping no conflict will occur
            key = str(keyword.replace(':', '_')) # no unicode
            criteria[key] = kw.get(keyword)
        counter = self._getCounter(**criteria)
        mapping.update(criteria)
        mapping['number'] = counter.hit()
        mapping['portal'] = getToolByName(self, 'portal_url').getPortalObject()
        self.logger.debug(mapping)
        return getEngine().getContext(mapping)


    security.declarePrivate('_getCounter')
    def _getCounter(self, **criteria):
        """Get counter for given keyword criteria
        """
        winner = None
        # stringify criteria in case it holds some integers
        scriteria = criteria.copy()
        for key, value in scriteria.items():
            scriteria[key] = stringify(value)
        for counter in self.objectValues():
            if scriteria == counter.getCriteria():
                winner = counter
                break
        if winner is None:
            # create it
            counter = self._createCounter(**criteria)
        return counter


    security.declarePrivate('_createCounter')
    def _createCounter(self, **criteria):
        """Create a counter using given keyword criteria
        """
        # create an id from criteria
        if not criteria:
            raise ValueError("cannot create counter with empty criteria")
        counter_id_list = []
        for key, value in criteria.items():
            counter_id_list.append("%s-%s"%(key, value))
        counter_id = '_'.join(counter_id_list)
        counter_id = generateId(counter_id, max_chars=100)
        counter = UidCounter(counter_id, self.counter_start, criteria)
        self._setObject(counter_id, counter)
        return self._getOb(counter_id)


InitializeClass(UidGenerator)
